dkg/frost/frost.go

```diff
@@ -0,0 +1,277 @@
+package frost
+
+import (
+	"encoding/json"
+
+	"github.com/bloxapp/ssv-spec/dkg"
+	"github.com/bloxapp/ssv-spec/types"
+	"github.com/coinbase/kryptology/pkg/core/curves"
+	"github.com/coinbase/kryptology/pkg/dkg/frost"
+	"github.com/coinbase/kryptology/pkg/sharing"
+	"github.com/herumi/bls-eth-go-binary/bls"
+	"github.com/pkg/errors"
+)
+
+var thisCurve = curves.BLS12381G1()
+
+type FROST struct {
+	identifier dkg.RequestID
+	network    dkg.Network
+	operatorID uint32
+	operators  []uint32
+	threshold  uint32
+	party      *frost.DkgParticipant
+
+	validatorPK    []byte
+	operatorShares map[uint32]*bls.SecretKey
+
+	round DKGRound
+	msgs  map[DKGRound]map[uint32]*ProtocolMsg
+}
+
+type DKGRound int
+
+const (
+	Preparation DKGRound = iota + 1
+	Round1
+	Round2
+	// Blame
+)
+
+func New(
+	requestID dkg.RequestID,
+	network dkg.Network,
+	i, t uint32,
+	operators []uint32,
+) dkg.KeyGenProtocol {
+
+	otherOperators := make([]uint32, 0)
+	for _, operatorID := range operators {
+		if operatorID == i {
+			continue
+		}
+		otherOperators = append(otherOperators, operatorID)
+	}
+
+	ctx := "string to prevent replay attacks"
+	party, _ := frost.NewDkgParticipant(i, uint32(len(operators)), ctx, thisCurve, otherOperators...)
+
+	return &FROST{
+		identifier: requestID,
+		network:    network,
+		operatorID: i,
+		threshold:  t,
+		operators:  operators,
+		party:      party,
+	}
+}
+
+func (fr *FROST) Start(init *dkg.Init) error {
+	return nil
+}
+
+func (fr *FROST) ProcessMsg(msg *dkg.SignedMessage) (bool, *dkg.KeyGenOutput, error) {
+	protocolMessage := &ProtocolMsg{}
+	if err := protocolMessage.Decode(msg.Message.Data); err != nil {
+		return false, nil, errors.Wrap(err, "could not decode protocol msg")
+	}
+
+	if fr.msgs[protocolMessage.Round] == nil {
+		fr.msgs[protocolMessage.Round] = make(map[uint32]*ProtocolMsg)
+	}
+
+	fr.msgs[protocolMessage.Round][uint32(msg.Signer)] = protocolMessage
+
+	switch protocolMessage.Round {
```

Comment:
We need to check what's the current round we are at. This means if we are at round2, we won't be handling round1 messages (which has already been handled in previous round).

Reply:
makes sense, I'll add the change